From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Peter Jones <pjones@redhat.com>
Date: Thu, 14 Mar 2024 16:58:09 -0400
Subject: [PATCH] configure.ac: do not preserve CFLAGS after setting up
 TARGET_CFLAGS

When we're building this in an rpm, RPM passes in certain
CFLAGS/CPPFLAGS/LDFLAGS/etc.  In the case of Fedora and RHEL, we do all
the x86 builds (x86_64-efi, i386-efi, and i386-pc) serially in separate
build directories, all on x86_64.  In that environment, rpm passes in
CFLAGS that are along the lines of:

    CFLAGS="${CFLAGS:--O2  -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=3 -Wp,-D_GLIBCXX_ASSERTIONS  -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -m64  -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer }" ; export CFLAGS ;

These are notably not appropriate for the i386-efi build.  configure
then uses that variable and other grub-specific information to build
TARGET_CFLAGS, HOST_CFLAGS, etc.

This would actually all be fine, except configure/automake/gentpl.py and
who knows what else play fast and loose with whether they're calling
things CFLAGS or HOST_CFLAGS or CFLAGS_PLATFORM or what, and sometimes
play swaparoo between foo_CFLAGS and CFLAGS for various reasons.

The result is on the i386-efi build, we sometimes get RPM's CFLAGS
included into TARGET_CFLAGS, which then causes multiple -march= and
-m(64|32) arguments.

This patch unsets the variables we care about once TARGET_CFLAGS and
HOST_CFLAGS are set up in configure.

Signed-off-by: Peter Jones <pjones@redhat.com>
---
 configure.ac | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/configure.ac b/configure.ac
index cd667a2eb75..de08f907957 100644
--- a/configure.ac
+++ b/configure.ac
@@ -2090,6 +2090,11 @@ HOST_CPPFLAGS="$HOST_CPPFLAGS -I\$(top_builddir)/include"
 TARGET_CPPFLAGS="$TARGET_CPPFLAGS -I\$(top_srcdir)/include"
 TARGET_CPPFLAGS="$TARGET_CPPFLAGS -I\$(top_builddir)/include"
 
+unset CFLAGS
+unset CPPFLAGS
+unset CCASFLAGS
+unset LDFLAGS
+
 GRUB_TARGET_CPU="${target_cpu}"
 GRUB_PLATFORM="${platform}"
 
